A notebook that creates an example of an SBOL2 Generic Location

PrashantVaidyanathan opened the pull request:
This addresses #26

jakebeal commented:
I think this isn't very informative yet, because it isn't concrete enough to really show the value of the representation. 
How about instead having this show a Cre-Lox recombination applied to activate a not-yet-selected coding sequence by un-reversing it? 

That would be a little more complicated, but I think a lot more compelling. The total representation would be:
- Two Component objects, for the Cre-Lox sites
- One SequenceAnnotation for the unknown CDS
- One Range (first recombination site) and two GenericLocation (CDS and second site)
- Two `SequenceConstraint` objects.

PrashantVaidyanathan commented:
@jakebeal - I agree with your comment there and I like the idea.
I have mixed feelings about including this example just for Generic Location because this has a lot more SBOL classes in it. 

How about we have a simpler notebook that just shows Generic Location. But we can also add another notebook showing a Cre-Lox recombination? I've added another notebook (outside the [CreatingSBOL2Objects](https://github.com/SynBioDex/SBOL-Notebooks/tree/Prashant/pysbol2-genericlocation/examples/sbol2/CreatingSBOL2Objects) folder). 

Happy to hear your thoughts.

PrashantVaidyanathan commented:
We can also add a note pointing users to the Cre-Lox example (or perhaps a ReadMe in the sbol2 folder)

jakebeal commented:
I think there's no point in using GenericLocation without this level of complexity.

I would thus suggest that instead of making a low-content GenericLocation, that we have this notebook's documentation point to the notebooks for SequenceConstraint, SequenceAnnotation, etc. as prerequisites

PrashantVaidyanathan commented:
I've modified the Generic Location notebook to reflect your comments. I've also added a link to the Cre-Lox notebook while providing users with a simpler code snippet to understand how to create a Generic Location.

jakebeal commented on examples/sbol2/CreLoxRecombination.ipynb:
```diff
@@ -0,0 +1,184 @@
+    "\n",
+    "Cre-Lox recombination is a site-specific recombination event, often used to modify sequences in synthetic biology by flipping or excising segments of DNA. This can be used to activate or deactivate genes or other elements.\n",
+    "\n",
+    "In this example, we will represent a situation where an unknown coding sequence (CDS) is initially in the wrong orientation and flanked by two loxP recombination sites. Upon recombination, the CDS will be flipped into the correct orientation. We'll use SBOL2 concepts like `Component`, `SequenceAnnotation`, `Range`, `GenericLocation`, and `SequenceConstraint` to model this event."
```
```suggestion
    "In this example, we will represent a situation where an unknown coding sequence (CDS) is initially in a reverse complement orientation (and thus not expressed) and flanked by two loxP recombination sites. Upon recombination, the CDS will be flipped into the forward orientation and become expressed. We'll use SBOL2 concepts like `Component`, `SequenceAnnotation`, `Range`, `GenericLocation`, and `SequenceConstraint` to model this event."
```

jakebeal commented on examples/sbol2/CreLoxRecombination.ipynb:
```diff
@@ -0,0 +1,184 @@
+    "\n",
+    "# Component for the second loxP site\n",
+    "loxP2 = sbol2.ComponentDefinition('loxP_site_2', sbol2.BIOPAX_DNA)\n",
+    "loxP2_seq = sbol2.Sequence('loxP_seq_2', 'ATAACTTCGTATAATGTATGCTATACGAAGTTAT', sbol2.SBOL_ENCODING_IUPAC)\n",
```
Since these two sites have the same sequence, we should use only one `ComponentDefinition` for `loxP`, and include two different instances via `Component` objects.

PrashantVaidyanathan replied:
Addressed in latest commit.

jakebeal commented on examples/sbol2/CreLoxRecombination.ipynb:
```diff
@@ -0,0 +1,184 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import sbol2\n",
```
This is a giant block of code. I would suggest breaking it into chunks with key code comments moving to instead by notebook text.

PrashantVaidyanathan replied:
Good point. Addressed in latest commit.

jakebeal commented on examples/sbol2/CreLoxRecombination.ipynb:
```diff
@@ -0,0 +1,184 @@
+    "doc.addComponentDefinition(loxP2)\n",
+    "\n",
+    "# Create a Component for the CDS (Coding Sequence)\n",
+    "# The CDS is unknown or \"not yet selected,\" so we can use a generic placeholder sequence\n",
```
We should not use any sequence at all. What you've got specified right here incorrectly constraints to an unknown of a specific length. Just omit the sequence.

PrashantVaidyanathan replied:
Addressed in latest commit.

jakebeal commented on examples/sbol2/CreLoxRecombination.ipynb:
```diff
@@ -0,0 +1,184 @@
+    "# Component for the first loxP site\n",
+    "loxP1 = sbol2.ComponentDefinition('loxP_site_1', sbol2.BIOPAX_DNA)\n",
+    "loxP1_seq = sbol2.Sequence('loxP_seq_1', 'ATAACTTCGTATAATGTATGCTATACGAAGTTAT', sbol2.SBOL_ENCODING_IUPAC)\n",
+    "loxP1.sequences = [loxP1_seq.identity]\n",
```
I think you should be able to do this more elegantly with something like this instead:
```suggestion
    "loxP1.sequences.add(loxP1_seq)\n",
```

jakebeal commented on examples/sbol2/CreLoxRecombination.ipynb:
```diff
@@ -0,0 +1,184 @@
+    "loxP2.sequences = [loxP2_seq.identity]\n",
+    "doc.addComponentDefinition(loxP2)\n",
+    "\n",
+    "# Create a Component for the CDS (Coding Sequence)\n",
```
```suggestion
    "# Create a ComponentDefinition for the CDS (Coding Sequence)\n",
```

A side note, which you might or might not want to say anything about: in SBOL3 this isn't required, because we can use a constraint on a SequenceFeature. It's so much nicer that way, because we don't have to have dummy TopLevel objects floating around.

jakebeal commented on examples/sbol2/CreLoxRecombination.ipynb:
```diff
@@ -0,0 +1,184 @@
+    "cds_annotation.locations.add(cds_location)\n",
+    "cds.sequenceAnnotations.add(cds_annotation)\n",
+    "\n",
+    "# GenericLocation for the second loxP site (second recombination site)\n",
```
```suggestion
    "# GenericLocation for the second loxP site (second recombination site) whose position is not known because it comes after the CDS\n",
```

jakebeal commented on examples/sbol2/CreLoxRecombination.ipynb:
```diff
@@ -0,0 +1,184 @@
+    "# Create SequenceConstraints to describe the relationships\n",
+    "# Constraint to ensure that loxP1 and CDS are adjacent\n",
+    "constraint1 = recombination_system.sequenceConstraints.create('loxP1_to_CDS')\n",
+    "constraint1.subject = loxP1_comp.identity\n",
```
I don't believe you need to pull identity out on any of these. I believe the assignment routine will do that automatically for you.

PrashantVaidyanathan replied:
Addressed in latest commit.

jakebeal commented on examples/sbol2/CreLoxRecombination.ipynb:
```diff
@@ -0,0 +1,184 @@
+    "\n",
+    "3. **Unknown CDS**: We defined an unknown or not-yet-selected coding sequence (CDS) using `GenericLocation`, which allows us to keep the boundaries flexible and represents the CDS in reverse complement orientation. The CDS is also added as a component to the system.\n",
+    "\n",
+    "4. **SequenceAnnotations**: We created `SequenceAnnotations` for each element (loxP sites and CDS), using a `Range` for the first loxP site and `GenericLocation` for the CDS and the second loxP site, enabling flexible positioning and orientation.\n",
```
```suggestion
    "4. **SequenceAnnotations**: We created `SequenceAnnotations` for each element (loxP sites and CDS), using a `Range` for the first loxP site and `GenericLocation` for the CDS and the second loxP site, in order to encode the orientations. If all of the `Component` objects were inline, the `GenericLocation` objects could instead be omitted.\n",
```

jakebeal commented on examples/sbol2/CreLoxRecombination.ipynb:
```diff
@@ -0,0 +1,184 @@
+    "   - The first loxP site precedes the CDS.\n",
+    "   - The CDS precedes the second loxP site.\n",
+    "\n",
+    "This setup simulates a Cre-Lox recombination event, where the recombination between the two loxP sites would flip the unknown CDS into the correct orientation, potentially activating it. By modeling the system this way, we maintain flexibility in defining the sequence, structure, and relationships between components."
```
```suggestion
    "This setup simulates a Cre-Lox recombination device, where the recombination between the two loxP sites would flip the unknown CDS into the correct orientation, potentially activating it. By modeling the system this way, we maintain flexibility in defining the sequence, structure, and relationships between components."
```

We aren't actually modeling the event here - for that, we'd need a `ModuleDefinition`, which is a pain to work with in SBOL2.

jakebeal reviewed (changes requested):
Some of my comments are "nice to haves", but I think it's important to avoid duplicating the loxP component and not add a sequence to the unknown CDS.

PrashantVaidyanathan commented:
> Some of my comments are "nice to haves", but I think it's important to avoid duplicating the loxP component and not add a sequence to the unknown CDS.

@jakebeal - I've addressed these comments in the latest commit. If you are happy with the changes, please feel free to merge this PR.

LukasBuecherl reviewed (approved):
Seems like Jake's comments have been addressed.
